### PR TITLE
[DRAFT] Add `series_library` to `PrimitiveBase`

### LINKED ISSUE #2165: Add series_library argument to transform primitives
- https://github.com/alteryx/featuretools/pull/2111

### PR DESCRIPTION
**Pull request by sbadithe**
- Fixes #2165 

The main purpose of this PR is to add a `series_library` attribute to `PrimitiveBase` and `FeatureBase`. In order to do this, it:  
- Adds `series_library` as an attribute on `primitive_dict` in the serialization and deserialization process 
- Adds branching logic to `feature_base` to determine the `series_library` based on the type of dataframe
- Refactors tests in `test_feature_serialization.py` and `test_features_serializer.py` to serialize the specific primitive instance on the feature, instead of just a new instance of the primitive
- Changes `Library` to be a string Enum to enable serialization

### CODE REVIEW

**Comment by dvreed77** [comment]
I'm trying to think of a good way to test this.  

We need to make sure series is set before we get the function in the `feature_set_calculator`

File: featuretools/feature_base/feature_base.py [hunk 1]
```diff
@@ -47,6 +51,13 @@ def __init__(
         # initialize if not already initialized
         if not isinstance(primitive, PrimitiveBase):
             primitive = primitive()
+
```
Comment (thehomebrewnerd) [inline featuretools/feature_base/feature_base.py:54]: What if the user has already instantiated the primitive before creating the Feature? I don't think the series type will get set in that case.

Comment (sbadithe) [reply featuretools/feature_base/feature_base.py:54]: If I am understanding correctly: this should be outside the `if-block`, right?

Comment (thehomebrewnerd) [reply featuretools/feature_base/feature_base.py:54]: Right - we always want to set the series type on the primitive, not just when it is instantiated during Feature creation.

File: featuretools/feature_base/feature_base.py [hunk 2]
```diff
@@ -47,6 +51,13 @@ def __init__(
...
+            # default library is PANDAS
+            if isinstance(dataframe, dd.DataFrame):
+                primitive.set_series_library(Library.DASK)
+            elif is_instance(dataframe, ps, "DataFrame"):
```
Comment (dvreed77) [inline featuretools/feature_base/feature_base.py:58]: I think we want to set the instance even if primitive is initialized

**Review by rwedge — COMMENTED** [review]
What if the primitive was instantiated before the feature was created?

Does the `series_library` information persist after using `ft.save_features` and `ft.load_features`?

File: featuretools/primitives/base/primitive_base.py [hunk 3]
```diff
@@ -156,3 +156,6 @@ def flatten_nested_input_types(input_types):
                 sub_input for input_obj in input_types for sub_input in input_obj
             ]
         return input_types
+
+    def set_series_library(self, series_library: str):
```
Comment (thehomebrewnerd) [inline featuretools/primitives/base/primitive_base.py:160]: Is there a reason to have this as a method rather than just setting the attribute directly?

**Comment by rwedge** [comment]
> We need to make sure series is set before we get the function in the `feature_set_calculator`

What if near the start of `calculate_feature_matrix` we looped through the features and ensured `series_library` was set to the right library type?

**Comment by sbadithe** [comment]
> What if the primitive was instantiated before the feature was created?
> 
> Does the `series_library` information persist after using `ft.save_features` and `ft.load_features`?

Could I test this by adding a test to `test_feature_serialization.py`?

**Comment by ozzieD** [comment]
It seems like you may have to add `series_library` to the serializer now.

File: setup.cfg [hunk 4]
```diff
@@ -44,7 +44,7 @@ include_package_data = True
 install_requires =
     click >= 7.0.0
     cloudpickle >= 1.5.0
-    dask[dataframe] >= 2021.10.0
+    dask[dataframe] == 2021.10.0
```
Comment (gsheni) [inline setup.cfg]:
```suggestion
    dask[dataframe] >= 2021.10.0, <2022.08.0
```

Comment (sbadithe) [reply setup.cfg]: thanks!

File: setup.cfg [hunk 5]
```diff
@@ -45,7 +45,7 @@ install_requires =
     click >= 7.0.0
     cloudpickle >= 1.5.0
     dask[dataframe] >= 2021.10.0, <2022.8.0
-    distributed >= 2021.10.0
+    distributed >= 2021.10.0, <2022.8.0
```
Comment (gsheni) [inline setup.cfg:48]: I added this.

File: featuretools/feature_base/features_deserializer.py [hunk 6]
```diff
@@ -140,6 +140,8 @@ def _deserialize_feature(self, feature_name):
 
         args = feature_dict["arguments"]
         feature = cls.from_dictionary(args, self.entityset, dependencies, primitive)
+        if "series_library" in feature_dict:
+            feature.primitive.series_library = feature_dict["series_library"]
```
Comment (dvreed77) [inline featuretools/feature_base/features_deserializer.py:144]: This works, but wondering if this is the best place to set series library on the primitivie

Comment (sbadithe) [reply featuretools/feature_base/features_deserializer.py:144]: As in, within `_deserialize_feature`?

**Comment by sbadithe** [comment]
Not sure what happened here...but I didn't intend to do that

**Pull request merged**